server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -538,47 +616,209 @@ public void trimUnreferencedReaders() throws IOException {
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    List<String> metadataFilesToBeDeleted = getMetadataFilesToBeDeleted(metadataFiles);
+
+                    // If no metadata file is filtered out, make sure to keep latest metadata file
+                    if (metadataFiles.size() == metadataFilesToBeDeleted.size() && metadataFiles.equals(metadataFilesToBeDeleted)) {
+                        metadataFilesToBeDeleted = metadataFiles.subList(1, metadataFiles.size());
+                    }
+
+                    if (metadataFilesToBeDeleted.isEmpty()) {
+                        logger.debug("No metadata files to delete");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    Set<Long> generationsToBeDeleted = getGenerationsToBeDeleted(metadataFilesToBeDeleted);
+                    if (generationsToBeDeleted.isEmpty() == false) {
+                        // Delete stale generations
+                        translogTransferManager.deleteGenerationAsync(
+                            primaryTermSupplier.getAsLong(),
+                            generationsToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Delete stale metadata files
+                        translogTransferManager.deleteMetadataFilesAsync(
+                            metadataFilesToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Delete stale primary terms
+                        deleteStaleRemotePrimaryTerms(metadataFiles);
+                    } else {
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                    }
+                } catch (Exception e) {
+                    remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                logger.error("Exception while listing translog metadata files", e);
             }
-            generationsToDelete.add(generation);
+        };
+        translogTransferManager.listTranslogMetadataFilesAsync(listMetadataFilesListener);
+    }
+
+    private Set<Long> getGenerationsToBeDeleted(List<String> metadataFilesToBeDeleted) throws IOException {
+        Set<Long> generationsToBeDeleted = new HashSet<>();
+
+        // From the remaining files, read oldest and latest
+        String oldestMetadataFileToBeDeleted = metadataFilesToBeDeleted.get(metadataFilesToBeDeleted.size() - 1);
+        String latestMetadataFileToBeDeleted = metadataFilesToBeDeleted.get(0);
+
```

Comment:
could this ordering be messed up for latest files during failover cases? or would primary term save us?

Reply:
These would still be ordered by primary term first